### PR TITLE
LPS-158646 Set Extended Properties in a JAX-RS Response Filter to ensure transactionality

### PR DESCRIPTION
**Pull request by ccorreagg**
This PR contains the code to ensure that all the operations executed in both: the JAX-RS resource and in the setExtendedProperties method, are executed over the same transaction.

Currently, the `TransactionContainerRequestFilter` class is in charge of creating and commiting/rollbacking every transaction inside a JAX-RS processing pipeline. As the setExtendedProperties was being executed after commiting/rollbacking the transaction, if there was any kind of error, the main operation executed ended successfully even if the problem happened.

To solve the issue, a new JAX-RS Container Response Filter class `EntityExtensionContainerResponseFilter` has been created, and its priority has been set to be executed **always before the code in charge of commiting/rollbacking the transaction**.

---

**Before the PR:**

1. `TransactionContainerRequestFilter`: It starts a new transaction
2. `BaseMessageBodyReader`: It checks if there are ExtensionProviders that could extend properties for the current operation.
3. The operation is executed (for example, to create a User Account).
4. `TransactionContainerRequestFilter`: It checks if the operation is successfully. In that case, the transaction is commited; otherwise is rollbacked.
5. `EntityExtensionWriterInterceptor`: The `setExtendedProperties` method is invoked. If there is any error, the transaction has been already commited, so it is not possible to rollback the main operation (for example, the User Account is already created).

**After the PR:**

1. `TransactionContainerRequestFilter`: It starts a new transaction
2. `BaseMessageBodyReader`: It checks if there are ExtensionProviders that could extend properties for the current operation.
3. The operation is executed (for example, to create a User Account).
4. `EntityExtensionContainerResponseFilter`: The `setExtendedProperties` method is invoked. If there is any error, the transaction has not been commited yet.
5. `TransactionContainerRequestFilter`: It checks if the operation is successfully. In that case, the transaction is commited; otherwise is rollbacked. If there was an error in the previous step, the transaction is rollbacked as well.
6. `EntityExtensionWriterInterceptor`: Only the `getExtendedProperties` method is invoked; the creating/updating operation has been successfully executed and commited.

---

**NOTE:** As the requests involves different JAX-RS Filters and Interceptors to be executed, an Integration Test has been created to ensure that transactionality keeps unaltered over the time.

### CODE REVIEW

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/container/response/filter/EntityExtensionContainerResponseFilter.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		if (clazz == null) {
+			return null;
+		}
+
+		EntityExtensionHandler entityExtensionHandler =
+			contextResolver.getContext(clazz);
+
+		if (entityExtensionHandler == null) {
+			entityExtensionHandler = _getEntityExtensionHandler(
+				clazz.getSuperclass(), contextResolver, mediaType);
```
Comment (javierdearcos) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/container/response/filter/EntityExtensionContainerResponseFilter.java:106]: Did you find any case where this is needed?

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/extension/test/EntityExtensionTest.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+
+/**
+ * @author Carlos Correa
+ */
+@RunWith(Arquillian.class)
+public class EntityExtensionTest {
```
Comment (javierdearcos) [inline modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/extension/test/EntityExtensionTest.java:74]: Very nice 👏

**Review by javierdearcos — APPROVED** [review]
Great job! 🥳

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/container/response/filter/EntityExtensionContainerResponseFilter.java (hunk 1 above)
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/container/response/filter/EntityExtensionContainerResponseFilter.java:106]: Yes, for example, for the case I am testing, the UserAccount DTO class, the value of `containerResponseContext.getEntityClass()` is `com.liferay.headless.admin.user.internal.dto.v1_0.converter.UserResourceDTOConverter$1` as the way the UserAccount is created is considered by Java as an "Inner class" of the DTO Converter `UserResourceDTOConverter`

image

**Comment by ccorreagg** [comment]
ci:forward

**Pull request closed without merging**